crates/worker/src/partition/snapshots/repository.rs

```diff
@@ -306,11 +322,133 @@ impl SnapshotRepository {
+            serde_json::from_slice(latest.bytes().await?.iter().as_slice())?;
+        trace!("Latest snapshot metadata: {:?}", latest);
+
+        let snapshot_metadata_path = object_store::path::Path::from(format!(
+            "{prefix}{partition_id}/{path}/metadata.json",
+            prefix = self.prefix,
+            partition_id = partition_id,
+            path = latest.path,
+        ));
+        let snapshot_metadata = self.object_store.get(&snapshot_metadata_path).await;
+
+        let snapshot_metadata = match snapshot_metadata {
+            Ok(result) => result,
+            Err(object_store::Error::NotFound { .. }) => {
+                info!("Latest snapshot points to a snapshot that was not found in the repository!");
+                return Ok(None); // arguably this could also be an error
+            }
+            Err(e) => return Err(e.into()),
+        };
+
+        let mut snapshot_metadata: PartitionSnapshotMetadata =
+            serde_json::from_slice(snapshot_metadata.bytes().await?.iter().as_slice())?;
+        if snapshot_metadata.version != SnapshotFormatVersion::V1 {
+            return Err(anyhow!(
+                "Unsupported snapshot format version: {:?}",
+                snapshot_metadata.version
+            ));
+        }
+
+        // The snapshot ingest directory should be on the same filesystem as the partition store
+        // to minimize IO and disk space usage during import.
+        let snapshot_dir = TempDir::with_prefix_in(
+            format!("{}-", snapshot_metadata.snapshot_id),
+            &self.base_dir,
+        )?;
+        debug!(
+            snapshot_id = %snapshot_metadata.snapshot_id,
+            path = ?snapshot_dir.path(),
+            "Getting snapshot data",
+        );
+
+        let directory = snapshot_dir.path().to_string_lossy().to_string();
+        let concurrency_limiter = Arc::new(Semaphore::new(DOWNLOAD_CONCURRENCY_LIMIT));
+        let mut downloads = JoinSet::new();
+        for file in &mut snapshot_metadata.files {
+            let filename = file.name.trim_start_matches("/");
+            let key = object_store::path::Path::from(format!(
+                "{prefix}{partition_id}/{path}/{filename}",
+                prefix = self.prefix,
+                partition_id = partition_id,
+                path = latest.path,
+                filename = filename,
+            ));
+            let file_path = snapshot_dir.path().join(filename);
+            let concurrency_limiter = Arc::clone(&concurrency_limiter);
+            let object_store = Arc::clone(&self.object_store);
+
+            downloads.spawn(async move {
+                let _permit = concurrency_limiter.acquire().await?;
+                let mut file_data = StreamReader::new(object_store.get(&key).await?.into_stream());
+                let mut snapshot_file = tokio::fs::File::create_new(&file_path).await?;
+                let size = io::copy(&mut file_data, &mut snapshot_file).await?;
```

Comment:
Maybe a sanity check that the downloaded file size matches what is expected from the snapshot metadata

Reply:
I want to do another pass and add some kind of checksum to the metadata file, but a size sanity check is a good and cheap start! Thanks!